### ISSUE TITLE
[css-a11y][css-display] display: contents; strips semantic role from elements 

### ISSUE THREAD

**Issue by jonnyjames** [w3c/csswg-drafts#3040]
The `display: contents;` property is part of the [CSS Display Module Level 3](https://www.w3.org/TR/css-display-3/). [APA tracking of CSS Display Module Level 3](https://www.w3.org/WAI/APA/wiki/CSS_Display_Module_Level_3)

`display: contents;` causes an element’s children to appear as if they were direct children of the element’s parent, ignoring the element itself. This can be useful when a wrapper element should be ignored when using CSS grid or similar layout techniques.

Unlike `display: none;` where users are unable to interact with the element to which it is applied, by using `display: contents;` the element is still visible and you can interact with it but only “through its contents” (See Ire Aderinokun’s blog post on “[How display: contents; Works](https://bitsofco.de/how-display-contents-works/)”). 

Scott O’Hara raised the issue that in the latest version of [Bikeshed](https://github.com/tabatkins/bikeshed), the use of `display: contents;` on the Table of Contents has led to accessibility issues for both screen readers and even standard keyboard tab use, which he notes in “[display contents causing TOC issues](https://github.com/tabatkins/bikeshed/issues/1316)". 

Hidde de Vries blog post “[More accessible markup with display: contents](https://hiddedevries.nl/en/blog/2018-04-21-more-accessible-markup-with-display-contents)” explains that current browsers that support `display: contents;` “do not only interpret it as a layout thing, but also derive meaning from it”. The display property only affects the visual layout, but by applying `display: contents;` to an element, browsers strip the element of its role semantics, causing accessibility issues. 

Currently, the note for [display properties](https://drafts.csswg.org/css-display/#the-display-properties) states that “The [display property](https://drafts.csswg.org/css-display/#propdef-display) has no effect on an element’s semantics: these are defined by the [document language](https://drafts.csswg.org/selectors-4/#document-language) and are not affected by CSS. Aside from the [`none`](https://drafts.csswg.org/css-display/#valdef-display-none) value, which also affects the aural/speech output [CSS-SPEECH-1](https://drafts.csswg.org/css-display/#biblio-css-speech-1)] and interactivity of an element and its descendants, the display property only affects visual layout: its purpose is to allow designers freedom to change the layout behavior of an element without affecting the underlying document semantics."
 
When `display: contents;` is applied to some elements, for example, form elements such as <select>, <input> and <textarea> (see [Effects of ‘display: contents;’ on Unusual Elements](https://drafts.csswg.org/css-display/#unbox-html)), it treats the `display: contents;` as `display: none;`.

Hidde de Vries has filed bugs with [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1455357), [Chrome 66](https://bugs.chromium.org/p/chromium/issues/detail?id=835455) and [Safari](https://bugreport.apple.com/web/?error&bugaccess) and [Webkit](https://bugs.webkit.org/show_bug.cgi?id=185679). Firefox 62 apparently has a fix that is included in the next release in early September 2018.  Pressure needs to be applied to the other browser vendors to fix the issues of `display: contents;` interfering with the element’s inherent role semantics.

The CSS A11Y Task Force recommends one or both of the following recommendations:

- Encouragement for browser vendors to fix this issue
- A warning note in the module document advising developers of the issues with `display: contents;`

**Comment by SelenIT** [comment]
Maybe the issue about focusing behavior of elements with `display:contents` (#2632) can be considered related here?

**Comment by fantasai** [comment]
Agreed this is a very unfortunate implementation bug. Happy to add a warning note until implementations fix their bugs.

**Comment by jensimmons** [comment]
Those of us teaching new layout techniques to the world have been talking about this for a while now — telling people they cannot use `display: contents;` until this is fixed. We changed Can I Use to reflect the problem: https://caniuse.com/#feat=css-display-contents to help spread the word.

**Comment by tabatkins** [comment]
The a11y-tree issue has been fixed in Chrome, thanks to some great work from Alice Boxhall. There's still potentially focus-related issues tho.

**Comment by jensimmons** [comment]
I believe this is already entirely fixed in Firefox, but someone on the call (I couldn't hear who) mentioned that he thought it was not yet fixed. If that's true — if something seems still not right in Firefox, do file a bug, and let me know.

**Comment by rachelandrew** [comment]
It was mentioned on the call that Chrome have a partial fix for this and are working on it. https://bugs.chromium.org/p/chromium/issues/detail?id=835455

Firefox has fixed some of the problems but issues still remain, see: https://bugzilla.mozilla.org/show_bug.cgi?id=1500958

**Comment by jensimmons** [comment]
Ticket for sorting out the current status of things on Can I Use: https://github.com/Fyrd/caniuse/pull/4706 

I believe all the partially-fixed stages (both the "still super broken" and the "almost there, but not") should have yellow-green boxes. (Which means Firefox drops back to yellow-green). There's a bit of debate about that. I'd love others to chime in. 

And there's a lot of detail to document around which browser having which level of support when & links to bug tickets. It'd be helpful if folks could help see that this is correct.

**Comment by jensimmons** [comment]
There is still more work for browser engines to do. It's documented at: https://caniuse.com/css-display-contents 

<img width="1452" alt="Screenshot 2023-10-06 at 11 50 54 AM" src="https://github.com/w3c/csswg-drafts/assets/108474/05d4560c-da0d-4c7e-9d67-d4fb6a5fab45">

**Comment by aardrian** [comment]
That visual Can I Use table implies all browsers have been broken in the same way.

All browsers ironed out the broader `display: contents` semantics issues many versions / years ago (barring a 113-114 Chrome regression). Safari has caught up as of v17 (with one known outstanding issue for headings: https://bugs.webkit.org/show_bug.cgi?id=261978).

Where all browsers are struggling is with interactive elements that have `display: contents`, including Safari (I filed a PR with Can I Use), owing to some disagreement on how to interpret the spec. As such, there is a request for position by Chrome from each engine:
* Mozilla: https://github.com/mozilla/standards-positions/issues/772
* WebKit: https://github.com/WebKit/standards-positions/issues/164

Which has lead to this WHATWG proposal:
https://github.com/whatwg/html/pull/9425

In other words, this may need to be coordinated with WHATWG as it appears to be running at the problem as well.

**Comment by aardrian** [comment]
@rthrejheytjyrtj545 I am not clear on what your link and comment mean for what I posted. Are you refuting, adding context, or something else?
